src/logic/errors.js

```diff
@@ -0,0 +1,77 @@
+import {createLogic} from 'redux-logic';
+import Analyzer from '../analyzers';
+import {getCurrentProject} from '../selectors';
+import {validatedSource} from '../actions/errors';
+import retryingFailedImports from '../util/retryingFailedImports';
+
+export async function importValidations() {
+  return retryingFailedImports(() =>
+    import(
+      /* webpackChunkName: "mainAsync" */
+      '../validations'
+    ),
+  );
+}
+
+const validateSource = async (
+  {language, source, projectAttributes},
+  dispatch,
+) => {
+  const validations = await importValidations();
+  const validate = validations[language];
+  const validationErrors = await validate(source, projectAttributes);
+  await dispatch(validatedSource(language, validationErrors));
+};
+
+const validateCurrentProject = createLogic({
+  type: [
+    'CHANGE_CURRENT_PROJECT',
+    'GIST_IMPORTED',
+    'SNAPSHOT_IMPORTED',
+    'PROJECT_RESTORED_FROM_LAST_SESSION',
+    'TOGGLE_LIBRARY',
+  ],
+  async process({getState}, dispatch, done) {
+    const state = getState();
+    const currentProject = getCurrentProject(state);
+    const analyzer = new Analyzer(currentProject);
+
+    const validatePromises = [];
+    for (const language of Reflect.ownKeys(currentProject.sources)) {
+      const source = currentProject.sources[language];
+      validatePromises.push(
+        validateSource(
+          {language, source, projectAttributes: analyzer},
+          dispatch,
+        ),
+      );
+    }
+    await Promise.all(validatePromises);
+    done();
+  },
+});
+
+const updateProjectSource = createLogic({
```

Comment:
I would actually probably call this `validateProjectOnChange` or something like that—generally I’d like to follow the principle of naming logics after what they actually do, rather than the action(s) they consume. (N.B. the sagas tend to follow the former naming convention, which is my fault!)